nucleus/core/bootstrap/src/main/java/com/sun/enterprise/glassfish/bootstrap/cfg/StartupContextUtil.java

```diff
@@ -36,46 +39,70 @@ public final class StartupContextUtil {
     private StartupContextUtil() {
     }
 
-    public static File getInstallRoot(Properties p) {
-        return absolutize(new File(p.getProperty(Constants.INSTALL_ROOT_PROP_NAME)));
 
+    /**
+     * @return autodetected glassfish directory based on where usually is this class.
+     */
+    public static File detectInstallRoot() {
+        // glassfish/modules/glassfish.jar
+        File bootstrapFile = findBootstrapFile();
+        // glassfish/
+        return bootstrapFile.getParentFile().getParentFile();
+    }
+
+
+    private static File findBootstrapFile() {
+        try {
+            return Which.jarFile(StartupContextUtil.class);
```

Comment:
In the original code used `ASMain.class`. Is not this a mistake or typo?

Reply:
It uses any class from the jar file to locate the jar file, so I used the closest one instead of making dependency cycle between packages. I will add javadoc for sure.